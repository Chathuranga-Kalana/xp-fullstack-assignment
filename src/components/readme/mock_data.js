export const username = "Chathuranga-Kalana";
export const repo_name = "Nunchi";
export const mockReadme = `# Nunchi

Nunchi is machine learning absenteeism hour prediction web application .
This web applicaton shows most reasons for why employees getting absent by showing graphs individual or overall.
It also generate PDF files

## prerequisite
- Python 3
- flask
- weasyprint(https://weasyprint.readthedocs.io/en/latest/install.html#gtk64installer)
- pandas
- numpy
- matplotlib
- seaborn`;
