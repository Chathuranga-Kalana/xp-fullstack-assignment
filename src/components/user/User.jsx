import React from "react";
import { Link } from "react-router-dom";

import { github_api } from "../../utils/api";
import "./user.css";

class User extends React.Component {
  constructor() {
    super();
    this.state = {
      reposDatas: [],
      networkResp: true,
      reposRespo: true
    };
  }
  handleSubmit(e) {
    e.preventDefault();
    // alert(this.state.userName);
    this.props.history.push(`/readme`);
  }
  onClick() {}

  componentDidMount() {
    const name = this.props.match.params.username;

    // alert(this.props.match.params.username);
    github_api(`/users/${name}`)
      .then(response => {
        this.setState({ networkResp: response.ok });
        return response.json();
      })
      .then(user => {
        this.setState({
          user: user
        });
      })
      .catch(error => {
        console.error(error);
        return { name: "Network Error" };
      });

    github_api(`/users/${name}/repos`)
      .then(response => {
        this.setState({
          reposRespo: response.ok
        });
        return response.json();
      })
      .then(json => {
        this.setState({
          reposDatas: json
        });
      })
      .catch(error => {
        console.error(error);
        return { name: "Network Error" };
      });
  }
  renderStat(stat) {
    return (
      <li key={stat.name} className="user-info__stat">
        <Link to={stat.html_url}>
          <p className="user-info__stat-value">{stat.value}</p>
          <p className="user-info__stat-name">{stat.name}</p>
        </Link>
      </li>
    );
  }

  render() {
    if (!this.state.user) {
      return <div className="user-page">LOADING...</div>;
    }
    const username = this.props.match.params.username;
    const user = this.state.user;
    //network responces
    const networkResp = this.state.networkResp;
    const reposRespo = this.state.reposRespo;

    // alert(networkResp);

    var { reposDatas } = this.state;

    // console.log(reposDatas);
    // User Overall Data
    const stats = [
      {
        name: "Public Repos",
        value: user.public_repos
      },
      {
        name: "Followers",
        value: user.followers
      },
      {
        name: "Following",
        value: user.following
      }
    ];

    return (
      <section>
        {networkResp ? (
          <div className="user-page">
            <div className="user-info">
              <a
                className="user-info__text"
                href={`https://github.com/${user.login}`}
              >
                <img
                  className="user-info__avatar"
                  src={user.avatar_url}
                  alt={`${user.login} avatar`}
                />
                <h2 className="user-info__title">
                  {user.login} ({user.name})
                </h2>
                <p className="user-info__bio">{user.bio}</p>
              </a>

              <ul className="user-info__stats">{stats.map(this.renderStat)}</ul>
            </div>
            <div>
              {reposRespo ? (
                <ul className="user-rep">
                  {reposDatas.map(reposData => (
                    <Link to={`/readme/${username}/${reposData.name}`}>
                      <li key={reposData.id}>{reposData.name}</li>
                    </Link>
                  ))}
                </ul>
              ) : (
                <p>No repositories Found</p>
              )}
            </div>
          </div>
        ) : (
          <p>User Not Found</p>
        )}
      </section>
    );
  }
}

export default User;
