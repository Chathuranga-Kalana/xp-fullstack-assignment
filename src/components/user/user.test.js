import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import User from "./User";
import { mockSuccessUser, mockSuccessUserRepos, username } from "./mock_data";

configure({ adapter: new Adapter() });

global.fetch = jest.fn().mockImplementation(path => {
  switch (path) {
    case `https://api.github.com/users/${username}`:
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockSuccessUser)
      });
    case `https://api.github.com/users/${username}/repos`:
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockSuccessUserRepos)
      });

    default:
      return {};
  }
});

describe("UserComponent", () => {
  it("should fetch data and render user component with user details and repository list", async () => {
    const user_comp = shallow(
      <User match={{ params: { username: username } }} />
    );

    await new Promise(setImmediate);
    user_comp.update();
    expect(user_comp.getElements()).toMatchSnapshot();
  });
});
