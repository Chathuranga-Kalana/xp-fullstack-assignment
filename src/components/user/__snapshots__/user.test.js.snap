// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserComponent should fetch data and render user component with user details and repository list 1`] = `
Array [
  <section>
    <div
      className="user-page"
    >
      <div
        className="user-info"
      >
        <a
          className="user-info__text"
          href="https://github.com/Chathuranga-Kalana"
        >
          <img
            alt="Chathuranga-Kalana avatar"
            className="user-info__avatar"
            src="https://avatars3.githubusercontent.com/u/25519689?v=4"
          />
          <h2
            className="user-info__title"
          >
            Chathuranga-Kalana
             (
            Chathuranga kalana
            )
          </h2>
          <p
            className="user-info__bio"
          />
        </a>
        <ul
          className="user-info__stats"
        >
          <li
            className="user-info__stat"
          >
            <ForwardRef>
              <p
                className="user-info__stat-value"
              >
                5
              </p>
              <p
                className="user-info__stat-name"
              >
                Public Repos
              </p>
            </ForwardRef>
          </li>
          <li
            className="user-info__stat"
          >
            <ForwardRef>
              <p
                className="user-info__stat-value"
              >
                1
              </p>
              <p
                className="user-info__stat-name"
              >
                Followers
              </p>
            </ForwardRef>
          </li>
          <li
            className="user-info__stat"
          >
            <ForwardRef>
              <p
                className="user-info__stat-value"
              >
                0
              </p>
              <p
                className="user-info__stat-name"
              >
                Following
              </p>
            </ForwardRef>
          </li>
        </ul>
      </div>
      <div>
        <ul
          className="user-rep"
        >
          <ForwardRef
            to="/readme/Chathuranga-Kalana/Nunchi"
          >
            <li>
              Nunchi
            </li>
          </ForwardRef>
          <ForwardRef
            to="/readme/Chathuranga-Kalana/pathFinder"
          >
            <li>
              pathFinder
            </li>
          </ForwardRef>
          <ForwardRef
            to="/readme/Chathuranga-Kalana/Personal-Assistant"
          >
            <li>
              Personal-Assistant
            </li>
          </ForwardRef>
          <ForwardRef
            to="/readme/Chathuranga-Kalana/SDGP"
          >
            <li>
              SDGP
            </li>
          </ForwardRef>
          <ForwardRef
            to="/readme/Chathuranga-Kalana/Slot-Machine"
          >
            <li>
              Slot-Machine
            </li>
          </ForwardRef>
        </ul>
      </div>
    </div>
  </section>,
]
`;
